File: mirror/tier2-php/showcase/type_constraints.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Type System
// Module: type_constraints.php

class ConstrainedValue {
    private $value;
    private $valueType;
    
    public function __construct($value, string $valueType) {
        if (gettype($value) !== $valueType) {
            throw new TypeError("Expected $valueType, got " . gettype($value));
        }
        $this->value = $value;
        $this->valueType = $valueType;
    }
    
    public function getValue() {
        return $this->value;
    }
}

class BoundedInt {
    private $value;
    private $min;
    private $max;
    
    public function __construct(int $value, int $minVal = 0, int $maxVal = 100) {
        if ($value < $minVal || $value > $maxVal) {
            throw new ValueError("Value $value outside bounds [$minVal, $maxVal]");
        }
        $this->value = $value;
        $this->min = $minVal;
        $this->max = $maxVal;
    }
    
    public function getValue(): int {
        return $this->value;
    }
}

function validate_number(int $n, int $minVal = -100, int $maxVal = 100): bool {
    return $n >= $minVal && $n <= $maxVal;
}

$result = [
    'constrained_int' => (new ConstrainedValue(42, 'integer'))->getValue(),
    'bounded_value' => (new BoundedInt(50, 0, 100))->getValue(),
    'validate_valid' => validate_number(50),
    'validate_invalid' => validate_number(200, 0, 100),
    'constraints_active' => true,
];

echo json_encode($result);
?>

File: mirror/tier2-dart/showcase/type_constraints.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Type System
// Module: type_constraints.dart

class ConstrainedValue<T> {
  final T value;
  final Type valueType;
  
  ConstrainedValue(this.value, this.valueType) {
    if (value.runtimeType != valueType) {
      throw TypeError();
    }
  }
  
  T getValue() => value;
}

class BoundedInt {
  final int value;
  final int min;
  final int max;
  
  BoundedInt(this.value, {this.min = 0, this.max = 100}) {
    if (value < min || value > max) {
      throw RangeError('Value out of bounds');
    }
  }
  
  int getValue() => value;
}

bool validateNumber(int n, {int minVal = -100, int maxVal = 100}) {
  return n >= minVal && n <= maxVal;
}

void main() {
  final result = <String, dynamic>{
    'constrainedInt': ConstrainedValue<int>(42, int).getValue(),
    'boundedValue': BoundedInt(50, min: 0, max: 100).getValue(),
    'validateValid': validateNumber(50),
    'validateInvalid': validateNumber(200, maxVal: 100),
    'constraintsActive': true,
  };
  print(result);
}
